Testnet Deployment

joshuakarp opened the pull request:
### Description
<!-- Write your description about what this PR is about. -->
Once MatrixAI/Polykey#310 has been merged, we'll finally be able to move onto the deployment of our testnet into AWS.

I foresee this to be achievable in a few stages:
1. Deploy Polykey onto AWS and ensure we have a working remote keynode (#194)
2. Create the testnet seed nodes and store these securely somewhere (#285)
3. Use these seed nodes for our domain name (#177)

* Relates to MatrixAI/Polykey#159 
* Relates to MatrixAI/Polykey#161 

### Issues Fixed

* ~~Fixes MatrixAI/Polykey#324~~ being completed in MatrixAI/Polykey#310

Testnet deployment:

* Fixes MatrixAI/Polykey#194
* Fixes MatrixAI/Polykey#285 
* Fixes MatrixAI/Polykey#177 

NodeGraph Structure:

* Relates MatrixAI/Polykey#244 
* Fixes MatrixAI/Polykey#317 
* Fixes MatrixAI/Polykey#158
* Fixes MatrixAI/Polykey#359 

Node Adding Policies:

* Fixes MatrixAI/Polykey#212 
* Fixes MatrixAI/Polykey#322
* Fixes MatrixAI/Polykey#344 
* Fixes MatrixAI/Polykey#363 

Node Removal Policies:

* Fixes MatrixAI/Polykey#150 
* Fixes MatrixAI/Polykey#226 
* Fixes MatrixAI/Polykey#345

### Tasks

- ~~[ ] 2. Complete MatrixAI/Polykey#324~~ - being completed in MatrixAI/Polykey#310
- [x] 1. Read through and revisit all previous issues
- [ ] 3.  Complete MatrixAI/Polykey#194
  * Ultimately provides us with a deployment of Polykey on AWS!
  * Task list there should be sufficient to follow process.
- [ ] 4.  Complete MatrixAI/Polykey#285 
- [ ] 5. Complete MatrixAI/Polykey#177 
- ~~[ ] 6. Created automated testing that utilises the testnet~~ - to be done in MatrixAI/Polykey#357 
  - ~~[ ] - MatrixAI/Polykey#159~~
  - ~~[ ] - MatrixAI/Polykey#161~~
- [x] 7. Add agent service `nodes` tests https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-1032173720
  * [x] - `nodesChainDataGet`
  * [x] - `nodesClosestLocalNode`
  * [x] - `nodesHolePunchMessage`
- [x] 8. Changed `Cmd` to `EntryPoint` because PK only has 1 executable and makes `docker run ...` easier.
- [x] 9. The `--seed-nodes='<defaults>;...'` will now mean that any specified seed nodes overrides the defaults rather than the other way around
- [x] 10. If seed nodes contains the own `NodeID`, the own `NodeId` will be automatically filtered.
- [x] 11. Make `pk agent start` return status information such as `nodeId` and not just `recoveryCode`
- [x] 12. All PK commands can contact a remote agent (except `agent start` and the like), it needs to be made more obvious whether we are contacting the local agent or remote agent https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1040120023
  - This is done by making the required client connection parameters of node ID, client host and client port all or nothing when set. This means if they are partially set, this results in a new exception `ErrorCLIClientOptions`.
  - If they are all set, that means a connection to an agent not specified by the local status.
  - If none are set, this means relying on the local status.
- [x] 13. The seed node should be adding the details of any node connecting to it to its own `NodeGraph`: see issue MatrixAI/Polykey#344
    - currently the seed node does **not** add an entry in its `NodeGraph` for a connecting node
    - this is essential for the Kademlia implementation to start to function, and for the node's details to eventually propagate out into the wider Polykey network
    - see https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1042518638
- [x] 14. Investigate and fix encountered Polykey errors/problems during testnet testing https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1041190329:
   - ~~[ ] a. No output provided on `pk identities trust`? However, I understand that we've previously shied away from providing "success" output on every single one of our commands. It can be tricky to diagnose problems when this is the case though.~~ - moved this to https://github.com/MatrixAI/js-polykey/issues/334#issuecomment-1043779027
   - ~~[ ] b. `pk identities trust` seemingly not adding a node to the gestalt graph when it doesn't already exist in `GestaltGraph`/`NodeGraph` - see https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1042543612 and https://matrixai.slack.com/archives/CEAUUV5QX/p1645069875382019~~ - moved this to https://github.com/MatrixAI/js-polykey/issues/334#issuecomment-1043779027
   - [x] c. `SyntaxError: Unexpected end of data` on an "invalid" node ID when using `decodeNodeId` - rebased on master has new js-id 3.3.1 that catches syntax errors when decoding multibase encoded strings
- [ ] 15. Refactor `NodeGraph` structure
  * [x] The new structure will use 3 nested sublevels to achieve MatrixAI/Polykey#244 
    - `buckets` sublevel will contain each bucket, where each bucket sublevel contains `NodeId` to `NodeData`
    - `meta` sublevel contains each bucket, where each bucket sublevel is a config structure, currently only `count`
    - `lastUpdated` or `index` sublevel that contains each bucket, and each bucket sublevel contains `lexi(lastUpdated)-NodeId` to `NodeId`, the key is a compound index, allowing us to efficiently acquire the most up-to-date or least up-to-date node entry
  * [x] New methods like `getNodes` and `getBuckets` will allow us to interrogate the state of the `NodeGraph` efficiently, by streaming data out of the `NodeGraph`, this will be important for debugging, and later analytics
  * [ ] Integration of `DBTransaction` and new iterator feature of sublevels that allow us to maintain a snapshot of the DB at a point in time
  * [x] Reviewing kademlia trie structure and bringing ideas in - reviewed and discovered that trie structure properties are not necessary in our NodeGraph, our design is already efficient and optimal
- [x] 16. Optimise `getClosestNode` method based on new `NodeGraph` structure
- [x] 17. Relationship between `NodeGraph` and `NodeManager` should be tightly coupled due to the adding and removal policies of nodes, maintaining the bucket limit must involving pinging old nodes

### Final checklist
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

* [ ] Domain specific tests
* [ ] Full tests
* [ ] Updated inline-comment documentation
* [ ] Lint fixed
* [ ] Squash and rebased
* [ ] Sanity check the final build

CMCDragonkai commented on src/validation/utils.ts:
```diff
@@ -173,6 +173,55 @@ function parsePort(data: any): Port {
+  }
+  return [seedNodeMappings, defaults];
+}
+
```
In #310 I noticed `NodeMapping` type is not quite correct. And I changed it to `Record<NodeIdEncoded, NodeAddress`. 

Since you've moved this here, I'm going to just copy this part into `validation/utils.ts` as well, and so when rebasing this, it will just need to be replaced by what already exists in #310 when it merges.

CMCDragonkai replied:
I've moved it into #310.

joshuakarp committed:
nodesGetAll command added: retrieves all buckets from the NodeGraph

CMCDragonkai committed:
WIP

- added getBuckets test for distance and lastUpdated order
- resetting buckets work
- changing utility names

tegefaulkes committed:
feat: added `NodeGraph.getClosestNodes()`

Implemented `getClosestNodes()` and relevant tests in `NodeGraph.test.ts`.

Relates to #212

tegefaulkes committed:
fix: `NodeManager.setNode` Properly handles adding new node when bucket is full

Logic of adding nodes has been split between `NodeManager` and `NodeGraph`. The `NodeGraph.setNode` just handles adding a node to the bucket where the `NodeManager.setNode` contains the logic of when to add the node

Relates #359

tegefaulkes committed:
test: fixed `nodeGraph` `get all buckets` test

fix into 7cc74d059f94739591307d9543fa73f999d3345f

tegefaulkes committed:
feat: added `connectionEstablishedCallback` to `Proxy`

Added a callback to the `Proxy` that is called when a `ForwardConnection` or `ReverseConnection` is established and authenticated. It is called with the following connection information; `remoteNodeId`, `remoteHost`, `remotePort` and `type`. They type signifies if it was a forward or reverse connection. Note that this is only triggered by composed connections.

Added a test for if the callback was called when a `ReverseConnection` is established.

Relates #332
Relates #344

tegefaulkes committed:
feat: `Proxy` trigger adding nodes to `Nodegraph`

Added an event to the `EventBus` that is triggered by the `Proxy`'s `connectionEstablishedCallback`. this adds the node to the `NodeGraph`.

Related #344

tegefaulkes committed:
feat: added optional timeout timers to `NodeConnectionManager` methods

In some cases we want to specify how long we attempt to connect to a node on a per-connection basis.

Related #363

tegefaulkes committed:
refactor: updated implementation of `nodePing`

Added `nodePing` command to `NodeConnectionManager` and `NodeManager.nodePing` calls that now. the new `nodePing` command essentially attempts to create a connection to the target node. In this process we authenticate the connection and check that the nodeIds match. If no address is provided it will default to trying to find the node through kademlia.

Related #322

tegefaulkes committed:
feat: `NodeManager.setNode` authenticates the added node

`setNode` now authenticates the node you are trying to add. Added a flag for skipping this authentication as well as a timeout timer for the authentication. this is shared between authentication new node and the old node if the bucket is full.

Related #322

tegefaulkes committed:
wip: general temp fixes to facilitate testing

tegefaulkes committed:
syntax: general linting

tegefaulkes committed:
fix: squash into ping node changes.

Updated `NodeConnectionManager.pingNode` to just use the proxy connection.

#322

tegefaulkes committed:
fix: setNode now does not ping the new node

#322

tegefaulkes committed:
feat: setNode concurrently pings multiple nodes

`setNode` now pings 3 nodes concurrently, updating ones that respond and removing ones that don't. If there is room in the bucket afterwards then we add the new node.

#322

tegefaulkes committed:
feat: nodeManager is now startStop

#322

tegefaulkes committed:
feat: async queueing for setting nodes

`setNode` now has a `blocking` flag that defaults to false. If it encounters a full bucket when adding a node then it will add the operation to the queue and asynchronously trys a blocking `setNode` in the background. `setNode`s will only be added to the queue if the bucket was full.

#322

tegefaulkes committed:
feat: establishing a `NodeConnection` adds the node to the nodeGraph

`NodeConnectionManager` now takes `NodeGraph` in  the `nodeConnectionManager.start` method. It has to be part of the start method since they are co-dependent. `NodeConnectionManager` cals `NodeManager.setNode()` when a connection is established. This fulfills the condition of adding a node to the graph during a forward connection.

Fixed up tests that were failing in relation to the `NodeManager` `StartStop` conversion.

#322

tegefaulkes committed:
wip: implementing `NodeManager.refreshBucket()`

#345

tegefaulkes committed:
feat: implemented `NodeManager.refreshBucket()`

This method preforms the kademlia `refreshBucket` operation. It selects a random node within the bucket and preforms a search for that node. The process exchanges node information with any nodes it connects to.

#345

tegefaulkes committed:
feat: implemented no activity timers and queuing for `refreshBucket`

Added queuing for `refreshBucket`. This means that buckets will be refreshed one at a time sequentially. This is to avoid doing a lot of costly refreshing all at once.

Added no activity for buckets. If a bucket hasn't been touched for a while, 1 hour by default, it will add a refresh bucket operation to the queue. Timers are disabled for buckets already in the queue. Only 1 timer is used for all buckets since only one of them can have the shortest timer and that's all we really care about.

#345

tegefaulkes committed:
feat: refreshing buckets when entering network

`nodeConnectionManager.syncNodeGraph` now refreshes all buckets above the closest node as per the kademlia spec. This means adding a lot of buckets to the refresh bucket queue when an agent is started.

#345

tegefaulkes committed:
feat: abort controller support for `NodeManager.refreshBucket`

Added support to cancel out of a `refreshBucket` operation. This is to allow faster stopping of the `NodeManager` by aborting out of a slow `refreshBucket` operation. This has been implemented with the `AbortController`/`AbortSignal` API. This is not fully supported by Node14 so we're using the `node-abort-controller` to provide functionality for now.

#345

emmacasolin commented on src/nodes/NodeManager.ts:
```diff
@@ -326,23 +379,131 @@ class NodeManager {
+        this.refreshBucketUpdateDeadline(bucketIndex);
+        return;
+      }
+      if (blocking) {
```
The logic here doesn't seem quite right to me - if `force` is set to true then, regardless of whether `blocking` is set or not, `nodeGraph.setNode()` is going to be called twice for the same node, right? Should it instead be `else if (blocking)`, or do we expect different behaviour if both `force` and `blocking` are set to true?

emmacasolin replied:
A diagram for the different paths that can be taken when calling `NodeManager.setNode()` would be helpful here - has one been made?

CMCDragonkai replied:
I don't have one atm. I believe there are 2 reasons for having a force vs block. Although this can be done in different ways.

1. Force means add the node to the node graph without checking if the node actually exists via a ping. This is often used when a ping has already occurred through one of the ways connections are made from the node or connections are made to the node. Or when the user explicitly wants to add the node for debugging purposes without caring about whether the node is alive.
2. Block is about choosing whether to run it by awaiting the result or by just putting it into the queue to run in the background. Sometimes you want the former and not the latter.

CMCDragonkai replied:
Note that I have not reviewed this deeply. But this was all based on the fact that there are multiple ways a connection may come into existence and we had to handle all these ways elegantly.

emmacasolin replied:
So it should be possible to specify both `force` and `blocking` and expect different behaviour? Because the current logic in `NodeManager.setNode()` is:

```
if (force):
  remove oldest node
  set new node immediately
if (blocking): // force can be true or false
  garbage collect old nodes
  set new node immediately if there's room // acts like force == false
else: // not blocking; force can be true or false
  queue set node
```

So if `force` is true then the node can be set twice. From your explanation it sounds like the logic should be more like:

```
if (blocking):
  garbage collect old nodes
  if (force && still no room):
    remove oldest node
  set new node immediately if there's room
else: // not blocking
  if (force):
    remove oldest node
    set new node immediately
  else:
    queue set node
```

Although in this case the `force` behaviour is essentially the same regardless of whether we're `blocking` or not, we're just performing extra (unnecessary) work in the former case. If we want to force a node to be set though then we always need to do that synchronously - if we add the operation to a queue then there's a chance that the bucket will be full again once we get around to it right? So I think just adding an `else` before `if (blocking)` is the most elegant solution.

```
if (force):
  force add synchronously
else if (blocking):
  garbage collect + add synchronously
else: // not blocking or force
  queue add asynchronously
```

CMCDragonkai replied:
That could be correct. Just check if it makes sense in the larger context of the rest of the node adding/removal policies.

emmacasolin replied:
From
> set a node, and if the bucket is full, drop the oldest node in the bucket (don't even use the concurrency parameter for the oldest) https://github.com/MatrixAI/js-polykey/issues/322#issuecomment-1082638907

and

> The setNode method has to deal with garbage collecting old nodes in a full bucket in case the force = false. https://github.com/MatrixAI/js-polykey/issues/322#issuecomment-1082760728

I believe that `else if (blocking)` is indeed the correct behaviour. I can fix that up.

emmacasolin committed:
fix: `setNode` no longer adds node twice when `force` is true
#322

emmacasolin committed:
feat: generic `SetNodeQueue` class for queuing `setNode` operations
`NodeManager.setNode` and `NodeConnectionManager.syncNodeGraph` now utilise a single, shared queue to asynchronously add nodes to the node graph without blocking the main loop. These methods are both blocking by default but can be made non-blocking by setting the `block` parameter to false.
#322

emmacasolin committed:
fix: `syncNodeGraph` during agent startup is now non-blocking
#322

emmacasolin committed:
syntax: renamed `SetNodeQueue` to `Queue`
#322

tegefaulkes committed:
refactor: cleaning up `Queue`

Just a small refactor. I've renamed some methods since queueStart and queuePush is unnecessarily verbose when Queue is its own class now. Also simplified some logic using the `promise` utility.

tegefaulkes commented on src/nodes/NodeConnectionManager.ts:
```diff
@@ -579,69 +582,102 @@ class NodeConnectionManager {
+        for (let i = bucketIndex; i < this.nodeGraph.nodeIdBits; i++) {
+          this.nodeManager?.refreshBucketQueueAdd(i);
+        }
       }
```
We don't really need to duplicate the code here. We can just create one lambda and use that,
```ts
const thing = async () => {
  // do the thing
}
if (!block) {
  this.queue.push(thing);
} else {
  await thing();
}
```

tegefaulkes committed:
tests: general fixes for failing tests

This contains fixes for failing tests as well as fixes for tests failing to exit when finished.

tegefaulkes committed:
syntax: added `@typescript-eslint/await-thenable` linting rule

This checks if we await things that are not promises. This is not a problem per se, but we generally don't want to await random things.

tegefaulkes committed:
fix: updated `@types/node-forge` version and fixed `keysUtils.getRandomBytes`

tegefaulkes committed:
test: added test to check if nodes are properly added to the seed nodes when entering the network

This tests for if the Seed node contains the new nodes when they are created. It also checks if the new nodes discover each other after being created.

Includes a change to `findNode`. It will no longer throw an error when failing to find the node. This will have to be thrown by the caller now. This was required by `refreshBucket` since it's very likely that we can't find the random node it is looking for.

tegefaulkes committed:
tests: Added agent service tests for `nodesChainDataGet`, `nodesClosestLocalNode` and `nodesHolePunchMessage`
